lib/AST/ASTContext.cpp

```diff
@@ -8194,6 +8204,12 @@ QualType ASTContext::mergeFunctionTypes(QualType lhs, QualType rhs,
         } else if (!rReturnAnnots.IsEmpty() && lReturnAnnots.IsEmpty()) {
           ReturnAnnots = rReturnAnnots;
           allLTypes = false;
+        } else if (lproto->isItypeGenericFunction() && !rproto->isItypeGenericFunction()) {
```

Comment:
I think you need to check that rproto is also not a generic function.  Generic function types should not be compatible with itype generic function type.  We will allow itype generic function types to be converted to generic types at assignments, call sites, and returns. We will not allow the reverse.

I think you should leave the code "as is" and open an issue to handle this after this PR is complete.    
This is already a large pull request.    You'll need to write some tests and alter some of the assignment compatibility code.  It's a special case that can wait.